### PR TITLE
Move QML rendering to a defined thread

### PR DESCRIPTION
**Pull request by zzmp**
A deadlock can be caused if the main thread tries to paint the scene, during a resize, whilst waiting for the "offscreen" QML to render. This occurs because the `OffscreenQMLSurface` allows any event to process while waiting for itself to render and retains its mutex.

This moves the `OffscreenQMLSurface`'s renderer to a separate thread, with its own message-passing queue, so that it no longer needs to allow all events to cycle while it waits.

### CODE REVIEW

**Comment by zzmp** [comment]
#### Testing

Test lots of QML surfaces, across lots of different display modes (e.g. switch between Oculus, Desktop, 3DTV, &c. a lot).

**Comment by hifi-gustavo** [comment]
Build console output for commit: f987da4c43eb96431a7b8d61521bf826fb0c5df0 available at: https://gist.github.com/0f281affd06a967ec5e5<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7399.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7399.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 62e635a52a02770efcd0799c50d261f9bdbca4dd available at: https://gist.github.com/0f281affd06a967ec5e5<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7399.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7399.dmg

**Comment by zzmp** [comment]
It looks like windows doesn't force you to forward declare a class within a cpp file.

**Comment by hifi-gustavo** [comment]
Build console output for commit: c4339dc85b9ace3850067b80cb32e5a9e2904110 available at: https://gist.github.com/0f281affd06a967ec5e5<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7399.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7399.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 41fcf4cc6d645d7261939939849cfa16efde6c6f available at: https://gist.github.com/0f281affd06a967ec5e5<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7399.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7399.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: d3d0f229aae2177157d8efd652e4f66654700c5b available at: https://gist.github.com/0f281affd06a967ec5e5<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7399.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7399.dmg

File: libraries/gl/src/gl/OffscreenQmlSurface.cpp [hunk 1]
```diff
@@ -60,268 +60,280 @@ static const QEvent::Type RENDER = QEvent::Type(QEvent::User + 2);
...
-        } else {
-            qDebug() << "QML render thread already completed";
-        }
+    setupFbo();
+    _canvas.doneCurrent();
+}
+
+void OffscreenQmlRenderThread::render() {
+    if (_surface->_paused) {
+        return;
```
Comment (jherico) [inline libraries/gl/src/gl/OffscreenQmlSurface.cpp]: You need to trigger the `_waitCondition` here or the main thread will hang.

File: libraries/gl/src/gl/OffscreenQmlSurface.cpp [hunk 2]
```diff
@@ -60,268 +60,280 @@ static const QEvent::Type RENDER = QEvent::Type(QEvent::User + 2);
...
+    QMutexLocker locker(&_mutex);
+    enqueue(new QEvent(type));
+    if (_isWaiting) {
+        _waitCondition.wakeOne();
     }
+}
 
+QEvent* OffscreenQmlRenderThread::Queue::take() {
+    QMutexLocker locker(&_mutex);
+    if (size() == 0) {
```
Comment (jherico) [inline libraries/gl/src/gl/OffscreenQmlSurface.cpp]: this should be a `while` loop, to guard against spurious wakeups.  also, prefer `isEmpty` to `size` when testing if the queue is empty.

File: libraries/gl/src/gl/OffscreenQmlSurface.cpp [hunk 3]
```diff
@@ -60,268 +60,280 @@ static const QEvent::Type RENDER = QEvent::Type(QEvent::User + 2);
...
-        }
-    }
+    virtual void run() override;
+    virtual bool event(QEvent *e) override;
 
-    void post(const QEvent::Type& type) {
-        QCoreApplication::postEvent(this, new QEvent(type));
-    }
+protected:
+    class Queue : public QQueue<QEvent*> {
```
Comment (jherico) [inline libraries/gl/src/gl/OffscreenQmlSurface.cpp]: I wouldn't recommend changing it for this PR, but in general, when there's a STL equivalent to a Qt type, prefer the STL version, i.e. `std::queue` instead of `QQueue` and `std::mutex` instead of `QMutex` below.  It will make things easier if we ever need to port away from Qt.

File: libraries/gl/src/gl/OffscreenQmlSurface.cpp (hunk 2 above)
Comment (zzmp) [inline libraries/gl/src/gl/OffscreenQmlSurface.cpp]: wow, huge oversight.

File: libraries/gl/src/gl/OffscreenQmlSurface.cpp (hunk 3 above)
Comment (zzmp) [inline libraries/gl/src/gl/OffscreenQmlSurface.cpp]: :+1:
I modeled this PR after the scenegraph render thread in Qt, so a lot of that came over (the `QQueue`, and `size() == 0`).

**Comment by hifi-gustavo** [comment]
Build console output for commit: cc83a9bc362dd7947ba12a93caf960b9a517e1ee available at: https://gist.github.com/0f281affd06a967ec5e5<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7399.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7399.dmg

**Comment by cozza13** [comment]
smoke tested and works
